infer/schema.go

```diff
@@ -126,13 +126,25 @@ func getResourceSchema[R, I, O any](isComponent bool) (schema.ResourceSpec, mult
 }
 
 func serializeTypeAsPropertyType(
-	t reflect.Type, indicatePlain bool, extType *introspect.ExplicitType,
+	t reflect.Type, indicatePlain bool, extType *introspect.ExplicitType, propType propertyType,
 ) (schema.TypeSpec, error) {
 	for t.Kind() == reflect.Pointer {
 		t = t.Elem()
 	}
+	// Provider authors should not be using resource.Asset directly, but rather types.AssetOrArchive.
+	// We will returrn an error if resource.Asset or resource.Archive is used directly for an input.
+	// pulumi/pulumi-go-provider#243
+	if propType == inputType && t == reflect.TypeOf(resource.Asset{}) {
+		return schema.TypeSpec{},
+			fmt.Errorf("resource.Asset is not a valid input type, please use types.AssetOrArchive instead")
+	}
+	if propType == inputType && t == reflect.TypeOf(resource.Archive{}) {
+		return schema.TypeSpec{},
+			fmt.Errorf("resource.Archive is not a valid input type, please use types.AssetOrArchive instead")
+	}
```

Comment:
AFAIK archives are correctly enforced, and are thus valid input types.

Under Pulumi's current type system, an asset is an archive but an archive is not an asset.